### PR TITLE
linter: added reports for static and non-public magic methods.

### LINKED ISSUE #551: Give a warning for non-public __call (and other)
## Code Example

```php
<?php

class A {
  private function __call($method, $args) {
    var_dump([$method => $args]);
  }
}

class B extends A {
}

$b = new B();
$b->dd(1);
```

## Actual Behavior

No warnings.

If we run that code, PHP will give a warning:
```
The magic method __call() must have public visibility and cannot be static
```

## Expected Behavior

Report that __call should be a public non-static method.

### PR DESCRIPTION
**Pull request by i582**
Added reports for static and non-public magic methods.
Added test.

Fixes #551

### CODE REVIEW

File: src/linter/root.go [hunk 1]
```diff
@@ -783,6 +783,39 @@ func (d *RootWalker) parseMethodModifiers(meth *stmt.ClassMethod) (res methodMod
 	return res
 }
 
+func (d *RootWalker) checkMagicMethod(meth node.Node, name string, modif methodModifiers) {
+	switch name {
+	case " __construct",
```
Comment (YuriyNasretdinov) [inline src/linter/root.go:788]: Constructors can be private and that's not an error

File: src/linter/report.go [hunk 2]
```diff
@@ -455,6 +455,20 @@ function performance_test() {}`,
...
+			Name:    "magicMethod",
+			Default: true,
+			Comment: `Report error in magic methods.`,
+			Before: `class Foo {
+  private function __call($method, $args) {} // The magic method __call() must have public visibility
+  static function __set($name, $value) {} // The magic method __set() cannot be static
+}`,
+			After: `class Foo {
+  public function __call($method, $args) {} // Ok
+  function __set($name, $value) {} // Ok
```
Comment (quasilyte) [inline src/linter/report.go:469]: It's generally a bad practice to omit access modifiers in classes. "public" is implied, but our examples shouldn't display a code that smells. :)

File: src/linter/report.go [hunk 3]
```diff
@@ -471,6 +471,20 @@ function performance_test() {}`,
 			Default: true,
 			Comment: `Report linter error.`,
 		},
+
+		{
+			Name:    "magicMethod",
```
Comment (quasilyte) [inline src/linter/report.go:476]: I would suggest calling it `magicMethodDecl` as it doesn't check all kinds of mistakes, but only declaration-related ones (visibility, instance/static modifiers).

File: src/linter/report.go [hunk 4]
```diff
@@ -471,6 +471,20 @@ function performance_test() {}`,
...
+			Default: true,
+			Comment: `Report error in magic methods.`,
```
Comment (quasilyte) [inline src/linter/report.go:478]: "Report issues in magic method declarations."

File: src/linter/root.go [hunk 5]
```diff
@@ -784,6 +784,53 @@ func (d *RootWalker) parseMethodModifiers(meth *stmt.ClassMethod) (res methodMod
...
+			d.Report(meth, LevelError, "magicMethod", "The magic method %s() cannot be static", name)
+		}
+
+	case "__callStatic":
+		// must be static and public
+		if !modif.static {
+			d.Report(meth, LevelError, "magicMethod", "The magic method %s() must be static", name)
+		}
+		if modif.accessLevel != meta.Public {
+			d.Report(meth, LevelError, "magicMethod", "The magic method %s() must have public visibility", name)
```
Comment (quasilyte) [inline src/linter/root.go:818]: Now I see that we haven't updated checker name in Report calls.

**Pull request merged**